frontend/app/services/session-service.server.ts

```diff
@@ -0,0 +1,89 @@
+ * the user's session.
+ *
+ * The SessionService module provides functions for creating, reading,
+ * updating, and deleting session data.
+ *
+ * Example usage:
+ *
+ *    const sessionStorage = await getSessionService().createSessionStorage();
+ *    const session = await sessionStorage.getSession(request.headers.get('Cookie'));
+ *    return json({}, {
+ *      headers: {
+ *        'Set-Cookie': await sessionStorage.commitSession(session)
+ *      }
+ *    });
+ *
+ * @see https://remix.run/docs/en/main/utils/sessions
+ */
+import { type Cookie, createCookie, createFileSessionStorage, createMemorySessionStorage, createSessionStorage } from '@remix-run/node';
+
+import { randomUUID } from 'node:crypto';
+
+import { getRedisService } from '~/services/redis-service.server';
+import { getEnv } from '~/utils/env.server';
+import { getLogger } from '~/utils/logging.server';
+
+const log = getLogger('session-service.server');
+
+export function getSessionService() {
+  const env = getEnv();
+
+  const sessionCookie = createCookie(env.SESSION_COOKIE_NAME, {
+    path: env.SESSION_COOKIE_PATH,
+    domain: env.SESSION_COOKIE_DOMAIN,
+    maxAge: env.SESSION_COOKIE_MAX_AGE,
+    secrets: [env.SESSION_COOKIE_SECRET],
+    secure: env.SESSION_COOKIE_SECURE,
+    httpOnly: env.SESSION_COOKIE_HTTP_ONLY,
+    // TODO :: GjB :: make these configurable?
+    sameSite: true,
+  });
+
+  async function createRedisSessionStorage({ cookie }: { cookie: Cookie }) {
+    const sessionId = randomUUID();
+    const redisService = getRedisService();
+    const setCommandOptions = { EX: env.SESSION_EXPIRES_SECONDS };
+
+    return createSessionStorage({
+      cookie,
+      createData: async (data) => {
+        log.debug(`Creating new session storage slot with id=[${sessionId}]`);
+        await redisService.set(sessionId, JSON.stringify(data), setCommandOptions);
+        return sessionId;
+      },
+      readData: async (id) => {
+        log.debug(`Reading session data for session id=[${id}]`);
+        return JSON.parse(await redisService.get(id));
+      },
+      updateData: async (id, data) => {
+        log.debug(`Updating session data for session id=[${id}]`);
+        await redisService.set(id, JSON.stringify(data), setCommandOptions);
+      },
+      deleteData: async (id) => {
+        log.debug(`Deleting all session data for session id=[${id}]`);
+        await redisService.del(id);
+      },
+    });
+  }
+
+  return {
+    createSessionStorage: () => {
+      switch (env.SESSION_STORAGE_TYPE) {
+        case 'memory':
+          // note: memory-backed sessions are not cleaned up; so use only during development!
+          return createMemorySessionStorage({ cookie: sessionCookie });
+        case 'file':
+          // note: memory-backed sessions are not cleaned up; so use only during development!
+          return createFileSessionStorage({ cookie: sessionCookie, dir: env.SESSION_FILE_DIR });
```

Comment:
**Nitpick:** This comment regarding in-memory sessions can be removed for file sessions

Reply:
That comment should be:

```
// note: file-backed sessions are not cleaned up; so use only during development!
```

Copy/paste error. :D (I'll fix)

Reply:
[Fixed](https://github.com/DTS-STN/canadian-dental-care-plan/pull/100/commits/c1d4a3c63d24c38a0c3d4472f3f5463aec365178#diff-0907994366d0437200280761d2f5478813c8b9e35737595580998eaf4ad7d7c0R81-R85)